File: Blad/Model/SearchIndex.swift
import Foundation

/// One markdown page with its contents, for searching and finding links.
nonisolated struct SearchEntry: Identifiable, Sendable {
    let url: URL
    let title: String
    /// Workspace and folders, e.g. "Notities › School".
    let location: String
    let text: String
    let modified: Date

    var id: URL { url }
}

nonisolated struct SearchResult: Identifiable, Sendable {
    let entry: SearchEntry
    /// The line around a match in the page's text; nil when the title matched.
    let snippet: String?

    var id: URL { entry.url }
}

nonisolated enum SearchIndex {
    private static let maxFileSize = 1_000_000
    private static let maxFiles = 5_000

    /// Reads every markdown page in the workspaces, newest first.
    /// Open pages use their text in memory, so unsaved edits are found too.
    static func build(workspaces: [URL], looseFiles: [URL], openTexts: [URL: String]) -> [SearchEntry] {
        var entries: [SearchEntry] = []
        var seen = Set<URL>()
        let keys: [URLResourceKey] = [.contentModificationDateKey, .fileSizeKey]

        func add(_ url: URL, location: String) {
            let url = url.standardizedFileURL
            guard entries.count < maxFiles, seen.insert(url).inserted else { return }
            let values = try? url.resourceValues(forKeys: Set(keys))
            guard (values?.fileSize ?? 0) <= maxFileSize,
                  let text = openTexts[url] ?? (try? String(contentsOf: url, encoding: .utf8)) else { return }
            entries.append(SearchEntry(
                url: url,
                title: url.deletingPathExtension().lastPathComponent,
                location: location,
                text: text,
                modified: values?.contentModificationDate ?? .distantPast
            ))
        }

        for workspace in workspaces {
            guard let enumerator = FileManager.default.enumerator(
                at: workspace,
                includingPropertiesForKeys: keys,
                options: [.skipsHiddenFiles, .skipsPackageDescendants]
            ) else { continue }

            let parentPath = workspace.deletingLastPathComponent().path
            while let url = enumerator.nextObject() as? URL {
                if FileNode.skippedFolders.contains(url.lastPathComponent) {
                    enumerator.skipDescendants()
                    continue
                }
                guard FileNode.isMarkdown(url) else { continue }
                let folders = url.deletingLastPathComponent().path.dropFirst(parentPath.count).split(separator: "/")
                add(url, location: folders.joined(separator: " › "))
            }
        }
        for url in looseFiles {
            add(url, location: url.deletingLastPathComponent().lastPathComponent)
        }
        return entries.sorted { $0.modified > $1.modified }
    }

    /// Title matches rank above matches in the text. An empty query lists recent pages.
    static func search(_ query: String, in entries: [SearchEntry]) -> [SearchResult] {
        let query = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else {
            return entries.prefix(10).map { SearchResult(entry: $0, snippet: nil) }
        }
        let options: String.CompareOptions = [.caseInsensitive, .diacriticInsensitive]

        var scored: [(result: SearchResult, score: Int)] = []
        for entry in entries {
            var score = 0
            if let range = entry.title.range(of: query, options: options) {
                score = range.lowerBound == entry.title.startIndex ? 300 : 200
            } else if query.count > 1, isSubsequence(query, of: entry.title) {
                score = 100
            }

            var snippet: String?
            if score == 0, let range = entry.text.range(of: query, options: options) {
                snippet = makeSnippet(entry.text, around: range)
                score = 50
            }
            if score > 0 {
                scored.append((SearchResult(entry: entry, snippet: snippet), score))
            }
        }

        return scored
            .sorted { $0.score != $1.score ? $0.score > $1.score : $0.result.entry.modified > $1.result.entry.modified }
            .prefix(50)
            .map(\.result)
    }

    /// "wkp" matches "Werkplan": the letters appear in order.
    private static func isSubsequence(_ query: String, of title: String) -> Bool {
        var remaining = Substring(query.lowercased().filter { !$0.isWhitespace })
        for character in title.lowercased() where character == remaining.first {
            remaining.removeFirst()
            if remaining.isEmpty { return true }
        }
        return remaining.isEmpty
    }

    private static func makeSnippet(_ text: String, around range: Range<String.Index>) -> String {
        let start = text.index(range.lowerBound, offsetBy: -40, limitedBy: text.startIndex) ?? text.startIndex
        let end = text.index(range.upperBound, offsetBy: 90, limitedBy: text.endIndex) ?? text.endIndex
        let excerpt = text[start..<end]
            .replacingOccurrences(of: "\n", with: " ")
            .trimmingCharacters(in: .whitespaces)
        return (start > text.startIndex ? "…" : "") + excerpt + (end < text.endIndex ? "…" : "")
    }
}

File: windows/Blad.Core/Search/SearchIndex.cs
using System.Globalization;
using Blad.Core.Pages;

namespace Blad.Core.Search;

/// <summary>One markdown page with its contents, for searching and finding links.</summary>
/// <param name="Location">Space and folders, e.g. "Wiskunde › Hoofdstuk 2".</param>
public sealed record SearchEntry(string Path, string Title, string Location, string Text, DateTime Modified);

/// <param name="Snippet">The text around a match inside the page; null when the title matched.</param>
public sealed record SearchResult(SearchEntry Entry, string? Snippet);

public static class SearchIndex
{
    private const long MaxFileSize = 1_000_000;
    private const int MaxFiles = 5_000;
    private const CompareOptions Loose = CompareOptions.IgnoreCase | CompareOptions.IgnoreNonSpace;

    /// <summary>Reads every page in the spaces, newest first. Open pages use their unsaved text.</summary>
    public static List<SearchEntry> Build(IEnumerable<string> spaces, IReadOnlyDictionary<string, string>? openTexts = null)
    {
        var entries = new List<SearchEntry>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var space in spaces)
        {
            var parent = Path.GetDirectoryName(Path.TrimEndingDirectorySeparator(space)) ?? space;
            foreach (var file in PageFiles.Enumerate(space))
            {
                if (entries.Count >= MaxFiles) break;
                if (!seen.Add(file)) continue;

                var info = new FileInfo(file);
                if (info.Length > MaxFileSize) continue;

                string text;
                if (openTexts is not null && openTexts.TryGetValue(file, out var open))
                {
                    text = open;
                }
                else
                {
                    try
                    {
                        text = File.ReadAllText(file).Replace("\r\n", "\n");
                    }
                    catch (IOException)
                    {
                        continue;
                    }
                    catch (UnauthorizedAccessException)
                    {
                        continue;
                    }
                }

                var folder = Path.GetRelativePath(parent, Path.GetDirectoryName(file) ?? parent);
                var location = string.Join(" › ", folder.Split([Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar], StringSplitOptions.RemoveEmptyEntries));
                entries.Add(new SearchEntry(file, Path.GetFileNameWithoutExtension(file), location, text, info.LastWriteTimeUtc));
            }
        }
        return entries.OrderByDescending(entry => entry.Modified).ToList();
    }

    /// <summary>Title matches rank above matches in the text. An empty query lists recent pages.</summary>
    public static IReadOnlyList<SearchResult> Search(string query, IReadOnlyList<SearchEntry> entries)
    {
        query = query.Trim();
        if (query.Length == 0) return entries.Take(10).Select(entry => new SearchResult(entry, null)).ToList();

        var compare = CultureInfo.InvariantCulture.CompareInfo;
        var scored = new List<(SearchResult Result, int Score)>();
        foreach (var entry in entries)
        {
            var score = 0;
            var titleIndex = compare.IndexOf(entry.Title, query, Loose);
            if (titleIndex >= 0) score = titleIndex == 0 ? 300 : 200;
            else if (query.Length > 1 && IsSubsequence(query, entry.Title)) score = 100;

            string? snippet = null;
            if (score == 0)
            {
                var index = compare.IndexOf(entry.Text, query, Loose, out var length);
                if (index >= 0)
                {
                    snippet = Snippet(entry.Text, index, length);
                    score = 50;
                }
            }
            if (score > 0) scored.Add((new SearchResult(entry, snippet), score));
        }

        return scored
            .OrderByDescending(item => item.Score)
            .ThenByDescending(item => item.Result.Entry.Modified)
            .Take(50)
            .Select(item => item.Result)
            .ToList();
    }

    /// <summary>"wkp" matches "Werkplan": the letters appear in order.</summary>
    private static bool IsSubsequence(string query, string title)
    {
        var letters = query.Where(c => !char.IsWhiteSpace(c)).Select(char.ToLowerInvariant).ToArray();
        var next = 0;
        foreach (var c in title.ToLowerInvariant())
        {
            if (next < letters.Length && c == letters[next]) next++;
        }
        return next == letters.Length;
    }

    private static string Snippet(string text, int index, int length)
    {
        var start = Math.Max(0, index - 40);
        var end = Math.Min(text.Length, index + length + 90);
        var excerpt = text[start..end].Replace('\n', ' ').Trim();
        return (start > 0 ? "…" : "") + excerpt + (end < text.Length ? "…" : "");
    }
}
